spec/java/util/ArrayList.SubList.lsl

```diff
@@ -0,0 +1,585 @@
+    }
+
+
+    @private constructor *.SubList (@target self: ArrayList_SubList, parent: ArrayList_SubList, fromIndex: int, toIndex: int)
+    {
+        // #problem: this constructor is useless
+        action NOT_IMPLEMENTED("inaccessible constructor");
+    }
+
+
+    // static methods
+
+    // methods
+
+    // within java.util.AbstractList
+    fun *.add (@target self: ArrayList_SubList, e: Object): boolean
+    {
+        action ASSUME(this.root != null);
+
+        val effectiveIndex: int = this.offset + this.length;
+        ArrayListAutomaton(this.root)._checkForComodification(modCount);
+        ArrayListAutomaton(this.root)._addElement(effectiveIndex, e);
+
+        _updateSizeAndModCount(+1);
+    }
+
+
+    fun *.add (@target self: ArrayList_SubList, index: int, element: Object): void
+    {
+        action ASSUME(this.root != null);
+
+        val effectiveIndex: int = this.offset + index;
+        ArrayListAutomaton(this.root)._rangeCheckForAdd(effectiveIndex);
+        ArrayListAutomaton(this.root)._checkForComodification(modCount);
+        ArrayListAutomaton(this.root)._addElement(effectiveIndex, element);
+
+        _updateSizeAndModCount(+1);
+    }
+
+
+    fun *.addAll (@target self: ArrayList_SubList, c: Collection): boolean
+    {
+        _addAllElements(this.length, c);
+    }
+
+
+    fun *.addAll (@target self: ArrayList_SubList, index: int, c: Collection): boolean
+    {
+        _addAllElements(index, c);
+    }
+
+
+    // within java.util.AbstractList
+    fun *.clear (@target self: ArrayList_SubList): void
+    {
+        action TODO();
+    }
+
+
+    fun *.contains (@target self: ArrayList_SubList, o: Object): boolean
+    {
+        result = _indexOfElement(o) >= 0;
+    }
+
+
+    // within java.util.AbstractCollection
+    fun *.containsAll (@target self: ArrayList_SubList, c: Collection): boolean
+    {
+        result = true;
+
+        if (!action CALL_METHOD(c, "isEmpty", []))
+        {
+            action ASSUME(this.root != null);
+
+            val rootStorage: list<Object> = ArrayListAutomaton(this.root).storage;
+            val end: int = this.offset + this.length;
+
+            if (c has ArrayList_SubListAutomaton)
+            {
+                val otherRoot: ArrayList = ArrayList_SubListAutomaton(c).root;
+                action ASSUME(otherRoot != null);
+
+                val otherStorage: list<Object> = ArrayListAutomaton(otherRoot).storage;
+                val otherOffset: int = ArrayList_SubListAutomaton(c).offset;
+                val otherEnd: int = otherOffset + ArrayList_SubListAutomaton(c).length;
+
+                action ASSUME(otherStorage != null);
+                action ASSUME(otherOffset >= 0);
+                action ASSUME(otherEnd >= 0);
+
+                var i: int = otherOffset;
+                action LOOP_WHILE(
+                    result && i < otherEnd,
+                    containsAll_loop_optimized(rootStorage, end, otherStorage, i, result)
```

Comment:
Ok. Do we need to force the USVM to make unnecessary range checks and state updates within the `Iterator` methods (it can be our own iterator automata) to make the specifications shorter?